// checkly/dist/constructs/alert-escalation-policy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlertEscalationBuilder = void 0;
// eslint-disable-next-line no-restricted-syntax
var AlertEscalationType;
(function (AlertEscalationType) {
    AlertEscalationType["RUN"] = "RUN_BASED";
    AlertEscalationType["TIME"] = "TIME_BASED";
})(AlertEscalationType || (AlertEscalationType = {}));
class AlertEscalationBuilder {
    static runBasedEscalation(failedRunThreshold, reminders, parallelRunFailureThreshold) {
        const options = {
            runBasedEscalation: {
                failedRunThreshold,
            },
            reminders,
            parallelRunFailureThreshold,
        };
        return this.alertEscalation(AlertEscalationType.RUN, options);
    }
    static timeBasedEscalation(minutesFailingThreshold, reminders, parallelRunFailureThreshold) {
        const options = {
            timeBasedEscalation: {
                minutesFailingThreshold,
            },
            reminders,
            parallelRunFailureThreshold,
        };
        return this.alertEscalation(AlertEscalationType.TIME, options);
    }
    static alertEscalation(escalationType, options) {
        var _a, _b, _c, _d;
        return {
            escalationType,
            runBasedEscalation: (_a = options.runBasedEscalation) !== null && _a !== void 0 ? _a : this.DEFAULT_RUN_BASED_ESCALATION,
            timeBasedEscalation: (_b = options.timeBasedEscalation) !== null && _b !== void 0 ? _b : this.DEFAULT_TIME_BASED_ESCALATION,
            reminders: (_c = options.reminders) !== null && _c !== void 0 ? _c : this.DEFAULT_REMINDERS,
            parallelRunFailureThreshold: (_d = options.parallelRunFailureThreshold) !== null && _d !== void 0 ? _d : this.DEFAULT_PARALLEL_RUN_FAILURE_THRESHOLD,
        };
    }
}
exports.AlertEscalationBuilder = AlertEscalationBuilder;
AlertEscalationBuilder.DEFAULT_RUN_BASED_ESCALATION = { failedRunThreshold: 1 };
AlertEscalationBuilder.DEFAULT_TIME_BASED_ESCALATION = { minutesFailingThreshold: 5 };
AlertEscalationBuilder.DEFAULT_REMINDERS = { amount: 0, interval: 5 };
AlertEscalationBuilder.DEFAULT_PARALLEL_RUN_FAILURE_THRESHOLD = { enabled: false, percentage: 10 };
