// checkly/dist/rest/environment-variables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class EnvironmentVariables {
    constructor(api) {
        this.api = api;
    }
    getAll() {
        return this.api.get('/v1/variables');
    }
    delete(environmentVariableKey) {
        return this.api.delete(`/v1/variables/${environmentVariableKey}`);
    }
    add(environmentVariableKey, environmentVariableValue, locked = false, secret = false) {
        return this.api.post('/v1/variables', { key: environmentVariableKey, value: environmentVariableValue, locked, secret });
    }
    get(environmentVariableKey) {
        return this.api.get(`/v1/variables/${environmentVariableKey}`);
    }
    // update environment variable with default locked value false
    update(environmentVariableKey, environmentVariableValue, locked = false, secret = false) {
        return this.api.put(`/v1/variables/${environmentVariableKey}`, { value: environmentVariableValue, locked, secret });
    }
}
exports.default = EnvironmentVariables;
